text/0000-dockerfiles.md

```diff
@@ -0,0 +1,218 @@
+version = "<extension version>"
+homepage = "<extension homepage>"
+description = "<extension description>"
+keywords = [ "<string>" ]
+
+[[extension.licenses]]
+type = "<string>"
+uri = "<uri>"
+```
+
+Extensions may be packaged and examined similar to buildpacks, but with analogous `pack extension` subcommands.
+
+Other pack CLI commands, such as `pack builder create`, will be extended to include support for extensions.
+
+Unlike buildpacks,
+- Extensions must not be included in a meta-buildpacks
+- Extensions must not have `order`/`group` definitions in `extension.toml`
+
+Extensions participate in the buildpack detection process, with the same UID, GID, and interface for `/bin/detect`.
+However,
+- `/bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.
+- Extensions may only output `provides` entries to the build plan. They must not output `requires`.
+- Extensions must all precede regular buildpacks in `order` definitions (e.g., in `builder.toml`).
+- Extensions are always `optional`.
+
+Extensions generate Dockerfiles before the regular buildpack build phase.
+To generate these Dockerfiles, the lifecycle executes the extension's `/bin/build` executable with the same UID, GID, and interface as regular buildpacks.
+However,
+- Extensions `/bin/build` must not write to the app directory.
+- Extensions `/bin/build` may be executed in parallel.
+- Extensions `<layers>` directory is replaced by an `<output>` directory.
+- If an extension is missing `/bin/build`, the extension root is treated as a pre-populated `<output>` directory.
+
+After `/bin/build` executes, the `<output>` directory may contain
+- `build.toml`, with the same contents as a normal buildpack's `build.toml`, but
+  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `build.Dockerfile` or `Dockerfile`
+- `launch.toml`, with the same contents as a normal buildpack's `launch.toml`, but
+  - Without the `processes` table array
+  - Without the `slices` table array
+  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `run.Dockerfile` or `Dockerfile`
+
+- Either `Dockerfile` or either or both of `build.Dockerfile` and `run.Dockerfile`
+
+Support for other instruction formats, e.g., LLB JSON files, could be added in the future.
+
+`build.Dockerfile`, `run.Dockerfile`, and `Dockerfile` target the builder image, runtime base image, or both base images, respectively.
+
+If no Dockerfiles are present, `/bin/build` may still consume build plan entries and add metadata to `build.toml`/`launch.toml`.
+
+Dockerfiles are applied to their corresponding base images after all extensions are executed and before any regular buildpacks are executed.
+Dockerfiles are applied in the order determined during buildpack detection.
```

Comment:
I have some questions about how multiple Dockerfiles are supposed to be chained together. 

Let's say after running extensions we have two build.Dockerfile's in this order:

```
ARG base_image
FROM ${base_image}

RUN echo "hello" > world.txt
```
```
ARG base_image
FROM ${base_image}

RUN cat world.txt > again.txt
```
What is the expected behavior when applying the second build.Dockerfile - should it succeed? If so then the `ARG` it receives is not the "original" build image but the intermediate image after applying the first build.Dockerfile, is that correct?

Assuming that's correct, and thinking about extending the build image "in container" with kaniko - we'd need to parse every build.Dockerfile that we receive and chain all the instructions together as if it were one big Dockerfile. There are some additional complexities with respect to multi-stage builds, but we haven't quite gotten to the point of thinking about those yet. It would be helpful though to get confirmation if we're on the right track.

Now let's say after running extensions we have two build.Dockerfile's in this order:

```
ARG base_image
FROM ${base_image}

RUN echo "hello" > world.txt
```
```
FROM alpine

RUN echo "hello" >> world.txt
```

How many "hello"s should exist in world.txt after both build.Dockerfiles are applied - just one? I'm assuming if we see a `FROM some-image` in any Dockerfile (not using the provided `base_image` arg), then all the Dockerfiles that come before it have no effect and should be ignored. Is that correct?

Reply:
Here are my thoughts...

For a `build` extension, the `base_image` would indeed be equal to "The current build image state" as the commands would be executed one after the other, in order of buildpack exts defined in `order.toml`. If we expect a clean slate after each `Dockerfile`, that would mean we need to run each `Dockerfile` in their own container and carry forward the snapshots to be re-assembled. That seems like it would be pretty slow and not worth the complexity IMO.

Your second example, I would think for a `build` extension you _must_ use `base_image` OR we need to always ignore the `FROM` for `Dockerfile.build` extensions and _just_ run the remaining instructions on the current state of the build container. The `Dockerfile.build` would essentially just be the transport for commands to run, The RFC has examples of using `Dockerfile` to create a new runtime base image, but I don't think we said you could change the build image base anywhere in the RFC.  The more I think about `build` extensions, the more I think we should _require_ that they are `FROM scratch`. Otherwise aren't we having kaniko executor pull/extract the base build layers as it processes N number of `Dockerfile.build` anyway? 

Now, as far as `Dockerfile.run` goes - this doesn't really work. If we intend on letting `Dockerfile.run` choose a `base_image` as described in the RFC...doesn't that mean we can really only support a single `Dockerfile.run` extension?

Imagine you have 3 `Dockerfile.run`s. One that calls `apt-get install x`, one that adds a file to `/opt` and one that changes the `base_image` to `scratch` from the previous target of `ubuntu:bionic`. If we executed them in that order, the `apt` installation would obviously not be valid as it would essentially get transplanted onto a scratch image. I guess we could have some restrictions around  `base_image` - saying it must be the same distro/version as the builder's target run-image...but that means `scratch` is off the table I think.

Similar to the first concern you had, I had _assumed_ all `Dockerfile.run` would be executed in the same container one after the other _on the target run image_. That doesn't work, however, if they all use different `base_image`. One, we don't know until execution what image we should be running on. We would need to run `kaniko/executor` and have it extract the `base_image` into the running container. We would also need to clean and extract the _next_ `base_image` to keep them isolated. Even then, we would end up making a resulting base image that would likely not function if they all used different `base_image`. They would also potentially have to duplicate work (`apt-get update`). I think sharing a container makes sense, but it means we need to know the base image ahead of time I think.

Reply:
Added https://github.com/buildpacks/rfcs/pull/173/files#r793096512 for clarity.